src/main/webapp/app/exercises/programming/shared/code-editor/ace/code-editor-ace.component.ts

```diff
@@ -208,8 +177,115 @@ export class CodeEditorAceComponent implements AfterViewInit, OnChanges, OnDestr
         if (this.annotationChange) {
             this.annotationChange.unsubscribe();
         }
-        if (this.resizeSubscription) {
-            this.resizeSubscription.unsubscribe();
+    }
+
+    /**
+     * Recalculates the position of annotations according to changes in the editor.
+     * Annotations are affected by changes in previous rows for row updates,
+     * in the same row and previous columns for column updates.
+     * @param change
+     */
+    updateAnnotations(change: TextChange) {
+        const {
+            start: { row: rowStart, column: columnStart },
+            end: { row: rowEnd, column: columnEnd },
+            action,
+        } = change;
+        if (action === 'remove' || action === 'insert') {
+            const sign = action === 'remove' ? -1 : 1;
+            const updateRowDiff = sign * (rowEnd - rowStart);
+            const updateColDiff = sign * (columnEnd - columnStart);
+
+            this.annotationsArray = this.annotationsArray.map((a) => {
+                return this.selectedFile !== a.fileName
+                    ? a
+                    : {
+                          ...a,
+                          row: a.row > rowStart ? a.row + updateRowDiff : a.row,
+                          column: a.column > columnStart && a.row === rowStart && a.row === rowEnd ? a.column + updateColDiff : a.column,
+                      };
+            });
+            this.displayAnnotations();
+        }
+    }
+
+    /**
+     * Sets the annotations for the editor.
+     * Checks for each annotation whether an updated version exists in local storage.
+     * @param annotations The new annotations array
+     */
+    setAnnotations(annotations: Array<Annotation> = []) {
+        if (annotations.length > 0) {
+            const sessionAnnotations = this.loadAnnotations();
+            this.annotationsArray = annotations
+                .map((a) => ({ ...a, hash: a.fileName + a.row + a.column + a.text }))
```

Comment:
I think you could merge the two maps into one and save one loop

Reply:
done